File: scripts/validate.ts
#!/usr/bin/env -S npx tsx
import { readFileSync, readdirSync } from "node:fs";
import { dirname, join, relative } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const SKILLS = join(ROOT, "skills");
const NAME_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

function scalar(frontmatter: string, key: string): string | undefined {
  const escaped = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const match = frontmatter.match(new RegExp(`^${escaped}:\\s*(.+?)\\s*$`, "m"));
  return match?.[1]?.trim().replace(/^(["'])(.*)\1$/, "$2");
}

export function validate(path: string): string[] {
  const errors: string[] = [];
  const text = readFileSync(path, "utf8");
  if (!text.startsWith("---\n")) return ["frontmatter must begin at byte 0"];
  const end = text.indexOf("\n---\n", 4);
  if (end < 0) return ["frontmatter closing delimiter not found"];
  const frontmatter = text.slice(4, end);
  const body = text.slice(end + 5).trim();
  const name = scalar(frontmatter, "name");
  const description = scalar(frontmatter, "description");
  if (!name) errors.push("name is required");
  else if (!NAME_RE.test(name) || name.length > 64) errors.push("name must be <=64 lowercase alphanumeric/hyphen characters");
  else if (name !== dirname(path).split("/").at(-1)) errors.push(`name '${name}' must match directory '${dirname(path).split("/").at(-1)}'`);
  if (!description) errors.push("description is required");
  else if (description.length > 1024) errors.push("description exceeds 1024 characters");
  if (!body) errors.push("body is required");
  if (text.split(/\r?\n/).length - (text.endsWith("\n") ? 1 : 0) > 500) errors.push("SKILL.md exceeds the recommended 500 lines");
  return errors;
}

export function main(): number {
  const paths = readdirSync(SKILLS, { withFileTypes: true }).filter((entry) => entry.isDirectory()).map((entry) => join(SKILLS, entry.name, "SKILL.md")).sort();
  if (paths.length === 0) { console.error("No skills found"); return 1; }
  let failed = false;
  for (const path of paths) {
    const errors = validate(path);
    if (errors.length) {
      failed = true;
      for (const error of errors) console.log(`FAIL ${relative(ROOT, path)}: ${error}`);
    } else console.log(`OK   ${relative(ROOT, path)}`);
  }
  console.log(`Validated ${paths.length} skill(s)`);
  return Number(failed);
}

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) process.exitCode = main();

File: scripts/validate.py
#!/usr/bin/env python3
"""Validate this repository's Agent Skills without external dependencies."""
from pathlib import Path
import re
import sys

ROOT = Path(__file__).resolve().parents[1]
SKILLS = ROOT / "skills"
NAME_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")

def scalar(frontmatter: str, key: str):
    match = re.search(rf"(?m)^{re.escape(key)}:\s*(.+?)\s*$", frontmatter)
    if not match:
        return None
    return match.group(1).strip().strip('"').strip("'")

def validate(path: Path):
    errors = []
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---\n"):
        return ["frontmatter must begin at byte 0"]
    end = text.find("\n---\n", 4)
    if end < 0:
        return ["frontmatter closing delimiter not found"]
    fm = text[4:end]
    body = text[end + 5:].strip()
    name = scalar(fm, "name")
    description = scalar(fm, "description")
    if not name:
        errors.append("name is required")
    elif not NAME_RE.fullmatch(name) or len(name) > 64:
        errors.append("name must be <=64 lowercase alphanumeric/hyphen characters")
    elif name != path.parent.name:
        errors.append(f"name {name!r} must match directory {path.parent.name!r}")
    if not description:
        errors.append("description is required")
    elif len(description) > 1024:
        errors.append("description exceeds 1024 characters")
    if not body:
        errors.append("body is required")
    if len(text.splitlines()) > 500:
        errors.append("SKILL.md exceeds the recommended 500 lines")
    return errors

def main():
    paths = sorted(SKILLS.glob("*/SKILL.md"))
    if not paths:
        print("No skills found", file=sys.stderr)
        return 1
    failed = False
    for path in paths:
        errors = validate(path)
        if errors:
            failed = True
            for error in errors:
                print(f"FAIL {path.relative_to(ROOT)}: {error}")
        else:
            print(f"OK   {path.relative_to(ROOT)}")
    print(f"Validated {len(paths)} skill(s)")
    return int(failed)

if __name__ == "__main__":
    raise SystemExit(main())
